iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -324,6 +326,249 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
+      auto newRank =
+          shape.size() - batchingDims.size() - contractingDims.size() + 2;
+      llvm::SmallVector<int64_t, 4> newShape(newRank);
+      auto batchingSize = std::accumulate(
+          batchingDims.begin(), batchingDims.end(), 1,
+          [shape](const int64_t accum, const int64_t index) -> int64_t {
+            return accum * shape[index];
+          });
+      auto contractingSize = std::accumulate(
+          contractingDims.begin(), contractingDims.end(), 1,
+          [shape](const int64_t accum, const int64_t index) -> int64_t {
+            return accum * shape[index];
+          });
+
+      int parallelDimIndex, contractingDimIndex, parallelDimSize = 1;
+      if (contractingDims.front() - batchingDims.back() > 1) {
+        parallelDimIndex = 1;
+        contractingDimIndex = 2;
+        for (int i = batchingDims.back() + 1; i < contractingDims.front();
+             ++i) {
+          parallelDimSize *= shape[i];
+        }
+      } else {
+        contractingDimIndex = 1;
+        parallelDimIndex = 2;
+        for (int i = contractingDims.back() + 1; i < shape.size(); ++i) {
+          parallelDimSize *= shape[i];
+        }
+      }
+      newShape[0] = batchingSize;
+      newShape[contractingDimIndex] = contractingSize;
+      newShape[parallelDimIndex] = parallelDimSize;
+      return std::make_tuple(newShape, contractingDimIndex, parallelDimIndex);
+    };
+
+    int lhsContractingDimIndex, rhsContractingDimIndex, lhsParallelDimIndex,
+        rhsParallelDimIndex;
+    SmallVector<int64_t, 4> lhsNewShape, rhsNewShape;
+    std::tie(lhsNewShape, lhsContractingDimIndex, lhsParallelDimIndex) =
+        computeCollabsedShape(lhsShapeType.getShape(), lhsBatchingDims,
+                              lhsContractingDims);
+
+    std::tie(rhsNewShape, rhsContractingDimIndex, rhsParallelDimIndex) =
+        computeCollabsedShape(rhsShapeType.getShape(), rhsBatchingDims,
+                              rhsContractingDims);
+    SmallVector<int64_t, 4> resultNewShape = {lhsNewShape[0],
+                                              lhsNewShape[lhsParallelDimIndex],
+                                              rhsNewShape[rhsParallelDimIndex]};
+    Type dotGeneralResultType =
+        RankedTensorType::get(resultNewShape, resultType.getElementType());
+
+    auto loc = op.getLoc();
+    Value reshapedLhs = rewriter.create<mhlo::ReshapeOp>(
+        loc, RankedTensorType::get(lhsNewShape, lhsShapeType.getElementType()),
+        op.lhs());
+    Value reshapedRhs = rewriter.create<mhlo::ReshapeOp>(
+        loc, RankedTensorType::get(rhsNewShape, rhsShapeType.getElementType()),
+        op.rhs());
+
+    auto make1DElementsAttr = [&rewriter](ArrayRef<int64_t> integers) {
+      auto type = RankedTensorType::get({static_cast<int64_t>(integers.size())},
+                                        rewriter.getIntegerType(64));
+      return DenseIntElementsAttr::get(type, integers);
+    };
```

Comment:
Can we make it a static function since we also have the same lambda in other functions?